### ISSUE TITLE
Error: Cannot modify extensions in the VM when the VM is not running

### ISSUE THREAD

**Issue by philthynz** [hashicorp/terraform-provider-azurerm#6098]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version:

```
Terraform v0.12.21
+ provider.azurerm v2.0.0
+ provider.null v2.1.2
+ provider.random v2.2.1
```
### Affected Resource(s)

* `azurerm_virtual_machine_extension`

### Terraform Configuration Files

**vm.tf**
```
# Virtual Machine Extension to initialize Windows
resource "azurerm_virtual_machine_extension" "init_win_vm_extension" {
  count                = var.os_type == "Windows" ? var.instance_numbers : 0
  depends_on           = [azurerm_virtual_machine_data_disk_attachment.win_man_disk_att]
  name                 = "${var.team_name}_${var.release}_vm_ext${count.index}"
  virtual_machine_id   = element(azurerm_windows_virtual_machine.win_vm.*.id, count.index)
  publisher            = "Microsoft.Compute"
  type                 = "CustomScriptExtension"
  type_handler_version = "1.9"

  protected_settings = <<PROTECTED_SETTINGS
    {
      "commandToExecute": "powershell.exe -ExecutionPolicy Unrestricted -Command \"./init.ps1; exit 0;\""
    }
  PROTECTED_SETTINGS

  settings = <<SETTINGS
    {
        "fileUris": [
           "https://link"
        ]
    }
  SETTINGS
}
```

**create_vm_image.tf**
```
# Generalize the Windows VM
resource "null_resource" "generalize_win_vm" {
  for_each = var.create_vm_images ? var.vms_to_image : {}
  provisioner "remote-exec" {
    connection {
    type     = "winrm"
    https    = true
    insecure = true
    user     = var.vm_username
    password = random_password.vm_password.result
    host     = element(azurerm_windows_virtual_machine.win_vm.*.public_ip_address, each.value)
    use_ntlm = true
    }

    inline = [
      "C:\\Windows\\System32\\Sysprep\\sysprep.exe /generalize /oobe /shutdown"
    ]
  }
}
```

```
# Mark the Windows VM as generalized in Azure.
resource "null_resource" "az_generalize_win_vm" {
  for_each = var.create_vm_images ? var.vms_to_image : {}
  depends_on = [null_resource.generalize_win_vm]
  provisioner "local-exec" {
    command = <<EOT
    az vm deallocate --resource-group "${var.rc_grp}" --name ${element(azurerm_windows_virtual_machine.win_vm.*.name, each.value)}
    az vm generalize --resource-group "${var.rc_grp}" --name ${element(azurerm_windows_virtual_machine.win_vm.*.name, each.value)}
    EOT
  interpreter = ["PowerShell", "-Command"]
  }
}
```

### Panic Output

```
Error: compute.VirtualMachineExtensionsClient#Delete: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="OperationNotAllowed" Message="Cannot modify extensions in the VM when the VM is not running."
```

### Expected Behavior

The `azurerm_virtual_machine_extension` resources are deleted.

### Actual Behavior

The `azurerm_virtual_machine_extension` resources are not deleted because the machines have been shutdown, and marked as "deallocate" and "generalize" in order to image these VM's. After doing this the VM's can't be turned back on.

### Steps to Reproduce

`terraform apply`

### Important Factoids

I have some TF that will remotely sysprep a machine, shut it down and use az cli to "deallocate" and "generalize". After doing this the VM's are no longer running.

TF should force delete, or ignore the fact the VM's are not running.

**Comment by philthynz** [comment]
Work around is to use `terraform state rm module.vm.azurerm_virtual_machine_extension.init_win_vm_extension[0]`. I believe TF should have some logic to overcome this error.

**Comment by eodonoghue** [comment]
This is still and issue.. are there any plans to resolve it?

**Comment by NDavitt** [comment]
any plans to resolve this? Was it resolved in v1.0?

**Comment by tombuildsstuff** [comment]
👋 unfortunately this is an Azure API limitation - Terraform can't skip the deletion of the VM Extension here since we don't have the context to determine whether the VM is being deleted when deleting the VM extension itself - as such (unfortunately) this isn't something we can workaround, and the VM will need to be in a running state to delete this.

It's possible that we could look to conditionally start/stop the VM to add/remove the extension - and if that's something you'd like to see supported in the Provider then feel free to open a new Feature Request and we can take a look - but (to be honest) we'd generally recommend using `CustomData` against the Virtual Machine rather than VM Extensions if possible.

Since this is an API limitation I'm going to close this issue for the moment, but as mentioned above feel free to open a new Feature Request if you're interested in having support for the VM being (conditionally) started/stopped during the deletion of the VM Extension.

Thanks!
